lib/hydra/remote_identifier/remote_services/doi.rb

```diff
@@ -49,34 +50,90 @@ def call(payload)
 
       def uri_for_request(payload)
-        puts payload
         unless payload.fetch(:identifier_url).nil?
           uri_for_request = URI.parse(payload.fetch(:identifier_url))
         else
-          uri_for_request = URI.parse(File.join(url, 'shoulder', shoulder))
+          uri_for_request = URI.parse(File.join(url))
         end
         uri_for_request.user = username
         uri_for_request.password = password
         uri_for_request
       end
 
       def request(data, payload)
-        response = RestClient.post(uri_for_request(payload).to_s, data, content_type: 'text/plain', accept: 'text/plain')
-        matched_data = /\Asuccess:(.*)(?<doi>doi:[^\|]*)(.*)\Z/.match(response.body)
-        { identifier: matched_data[:doi].strip, identifier_url: doi_service_url(matched_data[:doi].strip) }
+        if payload[:identifier_url].nil?
+          # response = RestClient.post(@username + ":" + @url + "@", data, content_type: 'application/json', username: @username, password: @password)
+          datacite_resource = RestClient::Resource.new @url, @username, @password
+          response = datacite_resource.post data, content_type: 'application/json'
+        else
+          doi_id = JSON.parse(
+              RestClient::Request.execute method: :get, url: payload.fetch(:identifier_url), user: @username, password: @password
+            )["data"]["id"]
+
+          datacite_resource = RestClient::Resource.new @url + "/" + doi_id, @username, @password
+          response = datacite_resource.put data, content_type: 'application/json'
+        end
+        identifier = JSON.parse(response.body)["data"]["attributes"]["doi"]
+        identifier_url = @url + "/" + JSON.parse(response.body)["data"]["id"]
+        result = {"identifier": "doi:" + identifier,"identifier_url":identifier_url}
+
       rescue RestClient::Exception => e
         raise(RemoteServiceError.new(e, uri_for_request(payload), data))
       end
 
       def data_for_request(payload)
-        data = []
-        data << "_target: #{payload.fetch(:target)}"
-        data << "_status: #{payload.fetch(:status)}"
-        data << "datacite.creator: #{Array(payload.fetch(:creator)).join(', ')}"
-        data << "datacite.title: #{Array(payload.fetch(:title)).join('; ')}"
-        data << "datacite.publisher: #{Array(payload.fetch(:publisher)).join(', ')}"
-        data << "datacite.publicationyear: #{payload.fetch(:publicationyear)}"
-        data.join("\n")
+        #data << "_target: #{payload.fetch(:target)}"
+        #data << "_status: #{payload.fetch(:status)}"
+        #data << "_profile: #{payload.fetch(:profile)}"
+        #data << "datacite.creator: #{Array(payload.fetch(:creator)).join(', ')}"
+        #data << "datacite.title: #{Array(payload.fetch(:title)).join('; ')}"
+        #data << "datacite.publisher: #{Array(payload.fetch(:publisher)).join(', ')}"
+        #data << "datacite.publicationyear: #{payload.fetch(:publicationyear)}"
+        #data.join("\n")
```

Comment:
Can we remove all of the commented `#data` lines or is it good to keep them around for some reason?

Reply:
No reason to keep it around, I've removed it.